File: Chatbots/Python/Source/dataModel.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Tuple, Set, Optional, Literal

DeckPath = Tuple[str, ...]
CandidatePool = List["Card"]
TopicIndex = Dict[DeckPath, List["Card"]]
TokenSet = Set[str]
DocumentId = int
InvertedIndex = Dict[str, List[Tuple[DocumentId, int]]]
IdfMap = Dict[str, float]
SparseVector = Dict[str, float]
AlgorithmName = Literal["keyword", "tfidf"]


@dataclass(frozen=True)
class Card:
    """One Q-A item from a deck export."""
    guid: str
    deck_path: DeckPath
    question_raw: str
    answer_raw: str
    question_text: str
    answer_text: str
    tags: List[str]
    question_token_count: int


@dataclass(frozen=True)
class PreparedQuestion:
    """Cached features for keyword scoring."""
    guid: str
    token_set: TokenSet
    non_stopword_count: int
    question_token_count: int


@dataclass(frozen=True)
class TfidfIndex:
    """TF-IDF state for a candidate pool."""
    documents: List[Card]
    inverted_index: InvertedIndex
    idf: IdfMap
    document_norms: List[float]
    document_token_counts: List[int]


@dataclass(frozen=True)
class AnswerHit:
    """A ranked result."""
    guid: str
    score: float
    deck_path: DeckPath
    question_preview: Optional[str] = None


@dataclass(frozen=True)
class QueryRequest:
    """Resolved query from CLI."""
    topic: DeckPath
    algorithm: AlgorithmName
    k: int
    query_text: str
    query_id: Optional[str] = None


@dataclass(frozen=True)
class StageTimings:
    """Per-query timings (ms)."""
    parse_ms: float
    index_ms: float
    preprocess_ms: float
    rank_ms: float
    format_ms: float


@dataclass(frozen=True)
class LogRecord:
    """Structured log row for one answered query."""
    timestamp_iso: str
    language: Literal["python"]
    algorithm: AlgorithmName
    deck_size: int
    topic: str
    query_id: str
    query_text: str
    stage_ms: StageTimings
    wall_ms: float
    rss_kb: Optional[int]
    top: List[Tuple[str, float]]


@dataclass(frozen=True)
class InvalidRecord:
    """Rejected input line (for error logs)."""
    file_path: str
    line_number: int
    reason: str
    raw_line_preview: str


TIE_BREAKERS_ORDER: Tuple[str, ...] = (
    "more_non_stopword_overlaps",
    "shorter_candidate_question",
    "lexicographic_guid",
)


def deck_path_to_string(deck_path: DeckPath) -> str:
    """Join a DeckPath using '::'."""
    return "::".join(segment.strip() for segment in deck_path if segment.strip())


def string_to_deck_path(text: str) -> DeckPath:
    """Split 'A::B::C' into a DeckPath."""
    parts = [segment.strip() for segment in text.split("::")]
    return tuple(segment for segment in parts if segment)


def short_preview(text: str, max_length: int = 120) -> str:
    """Truncate with an ellipsis if over max_length."""
    trimmed = text.strip()
    if len(trimmed) <= max_length:
        return trimmed
    return trimmed[: max_length - 1] + "…"

File: Chatbots/Rust/Source/src/data_model.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;

/// A hierarchical deck path, e.g. ["Launch into Computing", "Unit 03 - Principles of Computer Science"].
pub type DeckPath = Vec<String>;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Card {
    pub guid: String,
    pub question_text: String, // normalised for matching (tags stripped, entities decoded, lowercased)
    pub answer_raw: String,    // raw HTML/text for display; never render without escaping
    pub deck_path: DeckPath,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AnswerHit {
    pub guid: String,
    pub score: f32,
    pub question_preview: Option<String>,
    pub deck_path: DeckPath,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QueryRequest {
    pub query_text: String,
    pub topic_text: String,
    pub algorithm_name: String,
    pub top_k: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StageTimings {
    pub parse_milliseconds: f64,
    pub index_milliseconds: f64,
    pub preprocess_milliseconds: f64,
    pub rank_milliseconds: f64,
    pub format_milliseconds: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogRecord {
    pub timestamp_iso: String,
    pub language: String,   // "rust" or "python"
    pub algorithm: String,  // "keyword" or "tfidf"
    pub deck_size: usize,
    pub topic: String,
    pub query_id: String,
    pub query_text: String,
    pub stage_milliseconds: StageTimings,
    pub wall_milliseconds: f64,
    pub rss_kilobytes: Option<u64>,
    pub top: Vec<(String, f64)>,
}

/// Invalid line metadata captured during deck parsing/validation.
#[derive(Debug, Clone)]
pub struct InvalidRecord {
    pub file_path: String,
    pub line_number: usize,
    pub reason: String,
    pub raw_line: String,
}

/// Convert a deck path to a human-readable string (uses "::" separator).
pub fn deck_path_to_string(deck_path: &DeckPath) -> String {
    deck_path.join("::")
}

/// Create a short preview of text, truncated to `max_length` with an ellipsis if needed.
#[allow(dead_code)]
pub fn short_preview(full_text: &str, max_length: usize) -> String {
    if full_text.len() <= max_length {
        full_text.to_string()
    } else if max_length <= 1 {
        "…".to_string()
    } else {
        let cut_index = max_length.saturating_sub(1);
        format!("{}…", &full_text[..cut_index])
    }
}

/// Convenience: build a GUID → Card index for fast lookups.
pub fn build_guid_index(cards: &[Card]) -> HashMap<String, Card> {
    let mut guid_index = HashMap::new();
    for card in cards {
        guid_index.insert(card.guid.clone(), card.clone());
    }
    guid_index
}

/// Optional helper for constructing `AnswerHit`.
pub fn make_hit(guid: String, deck_path: DeckPath, question_preview: Option<String>, score: f32) -> AnswerHit {
    AnswerHit {
        guid,
        deck_path,
        question_preview,
        score,
    }
}

/// Pretty print a hit line (useful during early testing).
impl fmt::Display for AnswerHit {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        if let Some(preview_text) = &self.question_preview {
            write!(
                formatter,
                "GUID={}  score={:.6}  topic={}  Q: {}",
                self.guid,
                self.score,
                deck_path_to_string(&self.deck_path),
                preview_text
            )
        } else {
            write!(
                formatter,
                "GUID={}  score={:.6}  topic={}",
                self.guid,
                self.score,
                deck_path_to_string(&self.deck_path)
            )
        }
    }
}
